third_party/move/move-compiler-v2/src/pipeline/reference_safety/mod.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, ty::ReferenceKind};
+use move_stackless_bytecode::function_target::FunctionTarget;
+use std::{collections::BTreeMap, rc::Rc};
+
+pub mod reference_safety_processor_v3;
+
+/// Annotation produced by implementations
+#[derive(Clone)]
+pub struct LifetimeAnnotation(pub BTreeMap<CodeOffset, LifetimeInfoAtCodeOffset>);
+
+impl LifetimeAnnotation {
+    /// Returns information for code offset.
+    pub fn get_info_at(&self, code_offset: CodeOffset) -> &LifetimeInfoAtCodeOffset {
+        self.0.get(&code_offset).expect("lifetime info")
+    }
+}
+
+/// Information present at each code offset
+#[derive(Clone)]
+pub struct LifetimeInfoAtCodeOffset {
+    /// Information hidden via a trait, specific to the processor.
+    before: Rc<dyn LifetimeInfo>,
+    after: Rc<dyn LifetimeInfo>,
+}
+
+impl Default for LifetimeInfoAtCodeOffset {
+    fn default() -> Self {
+        Self {
+            before: Rc::new(NoLifetimeInfo()),
+            after: Rc::new(NoLifetimeInfo()),
+        }
+    }
+}
+
+impl LifetimeInfoAtCodeOffset {
+    pub fn new(before: Rc<dyn LifetimeInfo>, after: Rc<dyn LifetimeInfo>) -> Self {
+        Self { before, after }
+    }
+
+    pub fn borrow_kind_before(&self, temp: TempIndex) -> Option<ReferenceKind> {
+        self.before.borrow_kind(temp)
+    }
+
+    pub fn borrow_kind_after(&self, temp: TempIndex) -> Option<ReferenceKind> {
+        self.after.borrow_kind(temp)
+    }
+}
+
+/// A trait to be implemented by reference safety processors
+pub trait LifetimeInfo {
+    fn borrow_kind(&self, temp: TempIndex) -> Option<ReferenceKind>;
+    fn display(&self, target: &FunctionTarget) -> Option<String>;
+}
+
+struct NoLifetimeInfo();
```

Comment:
Is this for code offsets that are not reachable?

Reply:
No, for being able to implement `Default`. Added a comment.